### PR TITLE
[Feature Branch] KV Cache Interface

### PR DESCRIPTION
**Pull request by dbogunowicz**
## Feature Preview

Feature branch that aggregates all the features constituting the KV Cache Interface implementation. This includes:

- [x] #1084 
- [x] #1089
- [x] #1073
- [x] #1108 

### No-cache inference
```python
from deepsparse import Pipeline
import time
start = time.time()
opt = Pipeline.create(task="opt",
                      model_path="/home/ubuntu/damian/sparseml/deployment",
                      engine_type = "onnxruntime",
                      max_generated_tokens=1)
prompt = "Who is the president of the United States?"
output = opt(sequences=prompt, return_logits=True)
```

```bash
sequences=['\n'] logits=array([[[-12.644863 , -12.9746065,   2.577626 , ..., -13.5366125,
         -13.376596 , -14.587112 ]]], dtype=float32) session_id=None # same as in pytorch inference
```
```bash
Ground truth: [-12.6449, -12.9746,   2.5776,  ..., -13.5366, -13.3766, -14.5871]
```
### Single-token engine decoding only:

```python
from deepsparse import Pipeline
opt = Pipeline.create(task="opt",
                      model_path="/home/ubuntu/damian/sparseml/deployment",
                      engine_type = "onnxruntime",
                      max_generated_tokens=128)
prompt = "Who is the president of the United States?"
output = opt(sequences=prompt)
print(output.sequences)
```

```bash
2023-06-27 07:55:20 deepsparse.transformers.engines.nl_decoder_engine INFO     Overwriting in-place the input shapes of the transformer model at /home/ubuntu/damian/sparseml/deployment/model.onnx
2023-06-27 07:55:24 deepsparse.utils.onnx INFO     Overwriting in-place the batch size of the model at /home/ubuntu/damian/sparseml/deployment/model.onnx
2023-06-27 07:56:37 deepsparse.transformers.engines.nl_decoder_engine INFO     Overwriting in-place the input shapes of the transformer model at /home/ubuntu/damian/sparseml/deployment/model.onnx
2023-06-27 07:56:40 deepsparse.utils.onnx INFO     Overwriting in-place the batch size of the model at /home/ubuntu/damian/sparseml/deployment/model.onnx
['\n\nThe president of the United States is the head of the executive branch of government. The president is the head of the executive branch of government, and the president is the head of the executive branch of government. The president is the head of the executive branch of government, and the president is the head of the executive branch of government.\n\nThe president is the head of the executive branch of government, and the president is the head of the executive branch of government. The president is the head of the executive branch of government, and the president is the head of the executive branch of government. The president is the head of the executive']
```
```bash
Ground truth: The president of the United States is the head of the executive branch of government. The president is the head of the executive branch of government, and the president is the head of the executive branch of government. The president is the head of the executive branch of government, and the president is the head of the executive branch of government.

The president is the head of the executive branch of government, and the president is the head of the executive branch of government. The president is the head of the executive branch of government, and the president is the head of the executive branch of government.
```

## Single-token engine and multi-token engine decoding:

```python
from deepsparse import Pipeline
opt = Pipeline.create(task="opt",
                      model_path="/home/ubuntu/damian/sparseml/deployment",
                      engine_type = "onnxruntime",
                      max_generated_tokens=128)
prompt = "Who is the president of the United States?" * 20
output = opt(sequences=prompt)
print(output.sequences)
```

```bash
2023-06-27 07:57:53 deepsparse.transformers.engines.nl_decoder_engine INFO     Overwriting in-place the input shapes of the transformer model at /home/ubuntu/damian/sparseml/deployment/model.onnx
2023-06-27 07:58:47 deepsparse.utils.onnx INFO     Overwriting in-place the batch size of the model at /home/ubuntu/damian/sparseml/deployment/model.onnx
2023-06-27 07:58:52 deepsparse.transformers.engines.nl_decoder_engine INFO     Overwriting in-place the input shapes of the transformer model at /home/ubuntu/damian/sparseml/deployment/model.onnx
2023-06-27 07:58:58 deepsparse.utils.onnx INFO     Overwriting in-place the batch size of the model at /home/ubuntu/damian/sparseml/deployment/model.onnx
['Who is the president of the United States?Who is the president of the United States?Who is the president of the United States?Who is the president of the United States?Who is the president of the United States?Who is the president of the United States?Who is the president of the United States?Who is the president of the United States?Who is the president of the United States?Who is the president of the United States?Who is the president of the United States?Who is the president of the United States?Who is the president of the United States?Who is the president of the United States?Who is']
```

```bash
Ground truth: Who is the president of the United States?Who is the president of the United States?Who is the president of the United States?Who is the president of the United States?Who is the president of the United States?Who is the president of the United States?Who is the president of the United States?
```

### Manual Tests
The script below terminates without raising an error
```python
from deepsparse import Pipeline



def test_pipeline(engine_type):
    opt = Pipeline.create(task="opt",
                          model_path="/home/ubuntu/damian/sparseml/deployment",
                          engine_type=engine_type,
                          max_generated_tokens=32)

    prompt1 = "Who is the president of the United States?"
    prompt2 = "Who is the president of the United States?" * 20

    # test correctness unbatched input for a single-token engine
    out = opt(sequences=prompt1)
    out_ = opt(sequences=[prompt1])
    for x in [out, out_]:
        assert x.sequences[0] == '\n\nThe president of the United States is the head of the executive branch of government. The president is the head of the executive branch of government, and the'

    # test correctness unbatched input for a multi-token engine (very long input)
    out = opt(sequences=prompt2)
    out_ = opt(sequences=[prompt2])
    for x in [out, out_]:
        assert x.sequences[0] == 'Who is the president of the United States?Who is the president of the United States?Who is the president of the United States?Who is the president of'

    # test correctness batched input same input lengths
    out = opt(sequences=[prompt1, prompt1])
    for x in range(2):
        assert out.sequences[x] == '\n\nThe president of the United States is the head of the executive branch of government. The president is the head of the executive branch of government, and the'

    # test correctness batched input different input lengths
    out = opt(sequences=[prompt1, prompt2])
    assert out.sequences[0] == '\n\nThe president of the United States is the head of the executive branch of government. The president is the head of the executive branch of government, and the'
    assert out.sequences[1] == 'Who is the president of the United States?Who is the president of the United States?Who is the president of the United States?Who is the president of'


test_pipeline(engine_type = "onnxruntime")
test_pipeline(engine_type = "deepsparse")
```

### Testing with `eval downstream`
```bash
python src/deepsparse/transformers/eval_downstream.py /home/ubuntu/damian/sparseml/deployment --dataset openai_humaneval --engine onnxruntime --max-samples 4
```

#### HF baseline:
```bash
{'perplexities': [6.606388092041016, 9.294904708862305, 17.560449600219727, 13.867135047912598], 'mean_perplexity': 11.832219362258911}
```
#### Result with kv cache model
```bash
openai_humaneval eval results: {'mean_perplexity': 11.834174752235413, 'perplexities': [6.607589244842529, 9.296476364135742, 17.56318473815918, 13.8694486618042]}
# results match the baseline
```
#### Result with non-kv cache model
```bash
openai_humaneval eval results: {'mean_perplexity': 11.834173798561096, 'perplexities': [6.607589244842529, 9.296476364135742, 17.563180923461914, 13.8694486618042]}
# results match the baseline
```

## Current Limitations
- we are unable to run the "autoregressive multi-token" inference scenario. To be added in the future
- due to the limited deepsparse engine for the kv cache, we are not using `LIB.kv_cache` object for cache manipulation. We are also unable to run multi-token inference in the engine due to the issue with "zero-length" cache ingestion. In the deepsparse engine inference, whenever  a sequence would normally be processed by the multi-token engine, the single-token engine will take over instead
- the initialization time for the deepsparse engine is long (a few minutes).

### CODE REVIEW

File: src/deepsparse/transformers/engines/nl_decoder_engine.py [hunk 1]
```diff
@@ -0,0 +1,336 @@
...
+    def __str__(self):
+        return f"{self.__class__.__name__}: {self.engine}"
+
+    def transfer_cache_state(self, cache: DecoderKVCache):
+        """
+        Transfers the kv cache state and the number of tokens processed
+        information from another NLDecoderEngine. Call this method when
+        you want to transfer the kv cache state from one engine to another.
+
+        :param cache: The `NLDecoderEngine` to transfer the kv cache state
```
Comment (dbogunowicz) [inline src/deepsparse/transformers/engines/nl_decoder_engine.py:168]:
```suggestion
        :param cache: The `DecoderKVCache` object to transfer to the engine
```

**Review by bfineran — APPROVED** [review]
LGTM - we 100% need a bit more testing, let's make a plan for that. Let's also include the deepsparse vs ort perplexities in the description

File: src/deepsparse/tasks.py [hunk 2]
```diff
@@ -95,6 +95,12 @@ class SupportedTasks:
         ),
     )
 
+    text_generation = namedtuple("text_generation", ["opt", "codegen", "bloom"])(
+        codegen=AliasedTask("codegen", []),
```
Comment (rahul-tuli) [inline src/deepsparse/tasks.py:99]: nit: add code-gen, code_gen? to aliases

File: src/deepsparse/transformers/engines/nl_decoder_engine.py [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+    ):
+
+        onnx_file_path, output_indices_to_be_cached = self.overwrite_onnx_model_inputs(
+            onnx_file_path=onnx_file_path,
+            batch_size=engine_args.get("batch_size", 1),
+            sequence_length=sequence_length,
+            input_ids_length=input_ids_length,
+        )
+        kv_cache_enabled = False
+        if sum(output_indices_to_be_cached):
```
Comment (rahul-tuli) [inline src/deepsparse/transformers/engines/nl_decoder_engine.py:77]: Why sum(output_indices_to_be_cached)

File: src/deepsparse/transformers/engines/nl_decoder_engine.py [hunk 4]
```diff
@@ -0,0 +1,336 @@
...
+            overwritten with the new input shapes
+        :param batch_size: The batch size to use for the input
+        :param sequence_length: The sequence length to use for the input
+        :param input_ids_length: The length of input_ids
+        :return: The path to the onnx model file that has been overwritten
+            with the new input shapes, as well as the indices of the inputs
+            that should be cached
+        """
+        model = onnx.load(onnx_file_path, load_external_data=False)
+        initializer_input_names = set([node.name for node in model.graph.initializer])
```
Comment (rahul-tuli) [inline src/deepsparse/transformers/engines/nl_decoder_engine.py:194]:
```suggestion
        initializer_input_names = set(node.name for node in model.graph.initializer)
```

File: src/deepsparse/transformers/engines/nl_decoder_engine.py [hunk 5]
```diff
@@ -0,0 +1,336 @@
...
+    def generate_token(self, logits: numpy.ndarray) -> numpy.ndarray:
+        """
+        Samples a token from the logits using the sampling temperature.
+        :param logits: the logits from the model with shape (vocab_size,)
+        :return: the sampled token
+        """
+        if self.deterministic:
+            return numpy.argmax(logits)
+
+        logits /= self.sampling_temperature
```
Comment (rahul-tuli) [inline src/deepsparse/transformers/engines/nl_decoder_engine.py:238]: will sampling temperature always be > 0; if yes could we add a pre-condition to the doctstring?

File: src/deepsparse/transformers/eval_downstream.py [hunk 6]
```diff
@@ -68,14 +68,45 @@
...
+        max_generated_tokens=1,
+    )
+    perplexity_metrics = Perplexity(pipeline=text_generation, batch_size=batch_size)
+    active_engines = [
+        engine
+        for engine in [text_generation.engine, text_generation.multitoken_engine]
+        if engine
+    ]
+    print("Engine info: ")
+    [print(f"{engine}\n") for engine in active_engines]
```
Comment (rahul-tuli) [inline src/deepsparse/transformers/eval_downstream.py:100]: nit: remove list comprehension from prints

**Review by rahul-tuli — DISMISSED** [review]

File: src/deepsparse/transformers/eval_downstream.py (hunk 6 above)
Comment (dbogunowicz) [reply src/deepsparse/transformers/eval_downstream.py:100]: why?

File: src/deepsparse/transformers/engines/nl_decoder_engine.py [hunk 7]
```diff
@@ -0,0 +1,336 @@
...
+        return [
+            name
+            for name in self.engine.input_names
+            if not name.startswith(_CACHE_INPUT_NAME)
+        ]
+
+    def __call__(
+        self,
+        inp: List[numpy.ndarray],
+        val_inp: bool = True,
```
[the remark this answers is not in the document]
Comment (dbogunowicz) [reply src/deepsparse/transformers/engines/nl_decoder_engine.py:129]: let's leave for consistency with other Engine-type objects.

File: src/deepsparse/transformers/engines/nl_decoder_engine.py (hunk 3 above)
Comment (dbogunowicz) [reply src/deepsparse/transformers/engines/nl_decoder_engine.py:77]: if sum(...) = 0, no outputs are cached so no kv cache detected. Else, at least one output is being cached.

File: src/deepsparse/transformers/README.md [hunk 8]
```diff
@@ -32,9 +33,9 @@ This grants the engine the flexibility to serve any model in a framework-agnosti
 
 The DeepSparse pipelines require the following files within a folder on the local server to properly load a Transformers model:
 - `model.onnx`: The exported Transformers model in the [ONNX format](https://github.com/onnx/onnx).
-- `tokenizer.json`: The [HuggingFace compatible tokenizer configuration](https://huggingface.co/docs/transformers/fast_tokenizers) used with the model.
```
[the remark this answers is not in the document]
Comment (dbogunowicz) [reply src/deepsparse/transformers/README.md:35]: Only sometimes, e.g. not in LLMs. This is optional, which is hinted in the rest of the README.md

File: src/deepsparse/transformers/engines/nl_decoder_engine.py [hunk 9]
```diff
@@ -0,0 +1,336 @@
...
+                    f"Unexpected external input name: {external_input.name}"
+                )
+
+        _LOGGER.info(
+            "Overwriting in-place the input shapes "
+            f"of the transformer model at {onnx_file_path}"
+        )
+        save_onnx(model, onnx_file_path)
+
+        output_indices_to_be_cached = [
```
[the remark this answers is not in the document]
Comment (dbogunowicz) [reply src/deepsparse/transformers/engines/nl_decoder_engine.py:223]: why?

**Review by bfineran — APPROVED** [review]

File: src/deepsparse/transformers/eval_downstream.py (hunk 6 above)
Comment (rahul-tuli) [reply src/deepsparse/transformers/eval_downstream.py:100]: Because it doesn't do anything? 🥲 and creates an unnecessary list; a simple for loop seems more apt

File: src/deepsparse/transformers/engines/nl_decoder_engine.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (rahul-tuli) [reply src/deepsparse/transformers/engines/nl_decoder_engine.py:226]: when I first read the variable name, I believe it should just contain a list of valid indices to cache, but it's actually more like a binary mask over the entire array?

**Review by rahul-tuli — APPROVED** [review]

**Review by bfineran — APPROVED** [review]
